1-Draft/RFCNNNN-Test-Connection.md

```diff
@@ -0,0 +1,125 @@
+---
+RFC: XXXX
+Author: Joel Sallow
+Status: Draft
+Version: 1.0
+Area: Cmdlets
+Comments Due: 6/10/19
+Plan to Implement: Yes
+---
+
+# Proposed Changes to `Test-Connection` Cmdlet
+
+## Syntax
+
+```
+# All Sets
+Test-Connection [-TargetName] <string[]> [-IPv4] [-IPv6] [-ResolveDestination] [-TimeoutSeconds <int>] [-Quiet] [<CommonParameters>]
+
+# Set 1
+[-Ping] [-Count <int>] [-Delay <int>] [-Source <string>] [-MaxHops <int>] [-BufferSize <int>] [-DontFragment]
+
+# Set 2
+[-Ping] [-Continues] [-Delay <int>] [-Source <string>] [-MaxHops <int>]  [-BufferSize <int>] [-DontFragment] 
+
+# Set 3
+Test-Connection [-TargetName] <string[]> -Traceroute [-Source <string>] [-MaxHops <int>] 
+
+# Set 4
+Test-Connection [-TargetName] <string[]> -MTUSizeDetect
+
+# Set 5
+Test-Connection [-TargetName] <string[]> -TCPPort <int> [-Source <string>]
+```
+
+## Changes
+
+1. All current host/information output is relegated to the `-Verbose` stream.
+2. No progress bar per committee decision in [PowerShell#6768](https://github.com/PowerShell/PowerShell/issues/6768).
+3. All output occurs as soon as data is available, individual records per ping reply / trace hop.
+4. All presented data should be readily available as properties to the user.
+
+## Output Type Proposed Members
+
+### PingStatus
+
+```csharp
+class PingStatus
+{
+    string Source { get; }
+    IPAddress Address { get; }
+    string Destination { get; }
+    int RoundtripTime { get; }
+    int BufferSize { get; }
+    PingOptions Options { get; }
+}
+
+// For -MtuSizeDetect switch
+class PingMtuDetect : PingStatus
```

Comment:
Did you consider just having MtuSize being a member of PingStatus but $null if `-DetectMtuSize` is not specified?

Reply:
We can do this, absolutely. However, that means that the default format view with the default behaviour will always have an empty column.

That strikes me as... Odd? Also, given the cmdlet's current behaviour with `-MtuSizeDetect` which causes the cmdlet to only send / retrieve a single PingReply (vs default behaviour W/O the switch being to send 4 pings instead) it seems a little better to have this as a separate object type imo.

Definitely open to alternate solutions though!

Reply:
You're reminded me that @powercode has a PR to only show properties of MeasureInfo objects that have values.  We can do the same thing here.

Reply:
Can you link me to that? I wasn't aware that the formatter supported this!

Reply:
https://github.com/PowerShell/PowerShell/pull/7104, this was merged but was reverted

Reply:
So we'd need to put at least those methods into the code base for that to happen. @powercode, would you be able to PR just those formatting methods in so that we can make use of them? I'm not familiar with why it was reverted, but I'm sure we could stand to use a feature like that in more places than just here. 🙂

Reply:
@SteveL-Msft I looked over these again -- I know you opened an issue for it in the PS repo -- and I realize we're going to hit an issue here; the reason this is fairly easily doable for MeasureInfo is that the format used is a List format, not a Table format. Implementing the same thing in a table format is... problematic at best.